envoy/api/v2/route/route.proto

```diff
@@ -99,6 +100,13 @@ message VirtualHost {
   // [#not-implemented-hide:]
   // Return a 401/403 when auth checks fail.
   auth.AuthAction auth = 9;
+
+  // The per_filter_config field can be used to provide virtual host-specific
+  // configurations for filters. The key should match the filter name, such as
+  // *envoy.buffer* for the HTTP buffer filter. Use of this field is filter
+  // specific; see the :ref:`HTTP filter documentation <config_http_filters>`
+  // for if and how it is utilized.
+  map<string, google.protobuf.Struct> per_filter_config = 12;
```

Comment:
Suggest to add another map:

// map of filter name to filter config sha of SHA(seriazlized_per_filter_config).
map<string, string>  per_filter_config_sha = 13;

Some filters need to pre-process the filter config.  This sha can be used to skip the pre-process step if fitler_config is not changed.

if (sha != old_saved_sha) {
   obj = pre_process_config(config_data);
}

Reply:
The benefit of using a separate map is:
only the filters wanted to have this optimization need to fill in the sha map.

Reply:
Even if we wanted to add the SHA map (and I'm not sure I agree it's worth the complexity), I'm not sure how it would be used unless we make the Envoy side feature a lot more complicated. Basically, since this information is delivered over RDS, it's going to be in the route table, and orthogonal from the filter. The filter will just be able to look up the proto and dynamic_cast it from the route. So I don't think any kind of SHA will really help here? What did you have in mind?

Reply:
I think this doesn't belong in the initial version. We can always add it later. Also, hashing isn't that expensive. The filter can just compute the `xxHash` or whatevs.

Reply:
Agreed. @qiwzhang can you open a new issue on the SHA stuff? Again I don't really understand how it would work at the route level in any case.

Reply:
I like to clarify how per_route_config_sha can be used.

A filter needs to preprocess the per_route_config.

 obj = pre_process(pre_route_config);

Such pre-processing can be slow, it should NOT be performed for each request. One way to optimize it is to save the preprocessed "obj" in some global config.  When a request comes,  the filter just needs to detect if per_route_config has been changed.  Without "sha" map,  the filter needs to serialize the proto and sha it to detect the config change.   If per-route config is big, these two steps can be slow too.

Another option is to add a new function in FilterFactory

void setRouteConfig(string unique_route_id,   Struct config) PURE;

This function will be called when RDS receives new route tables,  This will give filter chance to preprocess per_route config and save the pre-processed obj into its global map (not pre-request).

as   global_map[ route_id ] = obj;

During request,  obj can be retrieved as

 obj =  global_map[ filter_callback->route()->unique_route_id() ];

This is better than first approach, the pre-process is done in the config thread, not at worker thread.

Reply:
Created https://github.com/envoyproxy/envoy/issues/3008 to track this